examples/issue-628-subnets-for-each/main.tf

```diff
@@ -0,0 +1,57 @@
+provider "aws" {
+  region  = "us-east-1"
+  profile = "ng-operations"
+}
+
+module "vpc" {
+  source = "../../"
+
+  name = "for-each-example"
+
+  cidr = "10.0.0.0/16"
+
+  azs                 = ["us-east-1a", "us-east-1b", "us-east-1c"]
+  private_subnets     = ["10.0.0.0/24", "10.0.1.0/24", "10.0.2.0/24"]
+  database_subnets    = ["10.0.6.0/24", "10.0.7.0/24", "10.0.8.0/24"]
+  redshift_subnets    = ["10.0.9.0/24", "10.0.10.0/24", "10.0.11.0/24"]
+  elasticache_subnets = ["10.0.12.0/24", "10.0.13.0/24", "10.0.14.0/24"]
+  intra_subnets       = ["10.0.15.0/24", "10.0.16.0/24", "10.0.17.0/24"]
+  public_subnets      = ["10.0.254.240/28", "10.0.254.224/28", "10.0.254.208/28"]
+
+  single_nat_gateway = true
+  enable_nat_gateway = true
+
+  enable_s3_endpoint       = true
+  enable_dynamodb_endpoint = true
+}
+
+resource "aws_ram_resource_share" "private_subnets" {
+  for_each                  = module.vpc.private_subnets_map
```

Comment:
You can also replace this for_each with:
```
for_each = { for k,v in module.vpc.public_subnets : k => v }
```

It will work as excepted (no need to do anything with outputs).

Reply:
I know this does not solve the issue of having type lists that can be solved if we use maps, but this PR is not moving the needle towards us using maps in both places (in variables and in outputs).

Reply:
@antonbabenko you cannot replace that for_each with what you mentioned.  The issue that I opened in #628 has the reasoning why.  You cannot do a for_each across a map whos keys are not known at plan time.  `module.vpc.public_subnets` is a list of subnet IDs, those IDs are not known at plan time unless the subnets have already been created, which makes them not suitable for use as the key in a map which is referenced in a for_each.

The workaround for this is to apply in two steps, first to apply the `module.vpc` to create the VPC and the subnets and then to apply the rest of the code (the subnets IDs would be known at plan time on the second run as they would already exist).

This PR resolves that issue by outputting the data from the module in the form of a map whos keys CAN be computed at plan time (because they are based off information that is not based on a return from the AWS API, but rather information that is supplied to terraform directly).  This removes the requirement to do a two-step apply where you first apply the VPC module to ensure that the information being used in the map keys is located in the state.  This is the problem that your proposed solution does not resolve.

Reply:
While I do agree that this does not fully resolve the issue of the module receiving information as a list and being based off counts (lists) for resource creation, it does resolve the issue of outputs not being maps.  This PR should be viewed as a "first step" and the format of these outputs could be maintained for compatibility pretty easily as the internals of the module are reworked to build the resources based on maps/for_each allowing for consumers of the module more time to migrate their code to be based off of these map based outputs and avoid a compatibility crunch when the module is eventually reworked.

Reply:
@antonbabenko did my statements above clarify what this PR resolves and why it is different than the issue that you are speaking about?  This functionality would greatly increase the usability of this module in some more advanced scenarios.

Reply:
I am not sure because I didn't have time to go very deep since I wrote my comment, run it, and it worked 3 days ago. It will require some time to get through this PR.